### PR TITLE
[1LP][RFR] Catalog items collections conversion

### PR DESCRIPTION
**Pull request by quarckster**
Purpose
=================

Catalog items collections conversion. Apart of that some refactoring has been done:

- all catalog items related stuff moved to a separate directory;
- each catalog item type has its own catalog item class;
- catalog bundles moved out from catalog_items module and converted to collections as well;
- ansible playbook catalog item drastically reduced the code;
- refactored affected tests and fixtures;
- some tests marked as skipped, they require additional refactoring by their authors.

{{pytest: -v -k "test_catalog_item or test_rhev_iso_servicecatalog or test_service_ansible_playbook_crud or test_pxe_servicecatalog" --long-running}}

### CODE REVIEW

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 32f02de79e2539926238575bae6ef986ec7795eb:

`cfme/services/catalogs/catalog_items/__init__.py`:
- :red_circle: Line 436:9: **N802** *function name should be lowercase*
- :red_circle: Line 478:9: **N802** *function name should be lowercase*
- :red_circle: Line 512:9: **N802** *function name should be lowercase*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d37d5f222f3570dbc087160c53f2bee26b15e4a1:

`cfme/rest/gen_data.py`:
- :bangbang: Line 9:1: **E999** *IndentationError: unexpected indent*
- :bangbang: Line 9:2: **E111** *indentation is not a multiple of four*
- :bangbang: Line 9:2: **E113** *unexpected indentation*

`cfme/services/catalogs/catalog_items/__init__.py`:
- :red_circle: Line 444:9: **N802** *function name should be lowercase*
- :red_circle: Line 463:21: **F821** *undefined name 'CatalogBundlesCollection'*
- :red_circle: Line 487:9: **N802** *function name should be lowercase*
- :red_circle: Line 521:9: **N802** *function name should be lowercase*

`cfme/tests/services/test_iso_service_catalogs.py`:
- :bangbang: Line 98:12: **E999** *SyntaxError: invalid syntax*

`cfme/tests/services/test_pxe_service_catalogs.py`:
- :bangbang: Line 97:36: **E999** *SyntaxError: invalid syntax*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 21321f6aaee8dcde818510d3310685b3b94babcb:

`cfme/services/catalogs/catalog_items/__init__.py`:
- :red_circle: Line 444:9: **N802** *function name should be lowercase*
- :red_circle: Line 486:9: **N802** *function name should be lowercase*
- :red_circle: Line 520:9: **N802** *function name should be lowercase*

`cfme/tests/services/test_iso_service_catalogs.py`:
- :red_circle: Line 96:12: **F821** *undefined name 'appliance'*
- :red_circle: Line 97:9: **F821** *undefined name 'appliance'*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ca33433f78f587241ee4060303c5d40e46d639ae:

`cfme/services/catalogs/catalog_items/__init__.py`:
- :red_circle: Line 444:9: **N802** *function name should be lowercase*
- :red_circle: Line 486:9: **N802** *function name should be lowercase*
- :red_circle: Line 520:9: **N802** *function name should be lowercase*

`cfme/tests/services/test_iso_service_catalogs.py`:
- :red_circle: Line 96:12: **F821** *undefined name 'appliance'*
- :red_circle: Line 97:9: **F821** *undefined name 'appliance'*

Hmmm
:scream: You might like to check the lint of your code :point_up: 


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit edccd023913e4441e3b14363579144352832d51c:

`cfme/services/catalogs/catalog_items/__init__.py`:
- :red_circle: Line 450:9: **N802** *function name should be lowercase*
- :red_circle: Line 504:9: **N802** *function name should be lowercase*
- :red_circle: Line 538:9: **N802** *function name should be lowercase*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by quarckster** [comment]
Failure of test_pxe_servicecatalog is not related to the PR

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 1]
```diff
@@ -0,0 +1,572 @@
...
+            wait_for(lambda: not self.save_button.disabled, timeout='10s', delay=0.2)
+
+
+class TabbedEditCatalogItemView(ServicesCatalogView):
+    save_button = Button('Save')
+    reset_button = Button('Reset')
+    cancel_button = Button('Cancel')
+
+    @View.nested
+    class BasicInfoTab(Tab):  # noqa
```
Comment (rlbabyuk) [inline cfme/services/catalogs/catalog_items/__init__.py]: Question: Shouldn't nested view classes be lower case?

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+
+    def update(self, updates):
+        view = navigate_to(self, 'Edit')
+        changed = view.fill(updates)
+        if changed:
+            view.save_button.click()
+        else:
+            view.cancel_button.click()
+        view = self.create_view(DetailsCatalogItemView, override=updates)
+        assert view.is_displayed
```
Comment (rlbabyuk) [inline cfme/services/catalogs/catalog_items/__init__.py]: Required: Please add view.flash.assert_no_error here

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+        else:
+            view.fill({'options': {'btn_text': 'btn_text',
+                                   'btn_hvr_text': button_name,
+                                   'select_dialog': self.dialog,
+                                   'btn_image': self.button_icon_name},
+                       'advanced': {'system_process': 'Request',
+                                    'request': 'InspectMe'}})
+        view.add_button.click()
+        view = self.create_view(DetailsCatalogItemView)
+        time.sleep(5)
```
Comment (rlbabyuk) [inline cfme/services/catalogs/catalog_items/__init__.py]: Required: replace this sleep with wait_for view

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 4]
```diff
@@ -0,0 +1,572 @@
...
+        """
+        cat_item = self.instantiate(catalog_item_class, *args, **kwargs)
+        view = navigate_to(cat_item, 'Add')
+        view.fill(cat_item.fill_dict)
+        view.add_button.click()
+        view = self.create_view(AllCatalogItemView)
+        # TODO move this assertion to tests
+        # view.flash.assert_success_message('Catalog Item "{}" was added'.format(
+        #     cat_item.name), partial=True)
+        assert view.is_displayed
```
Comment (rlbabyuk) [inline cfme/services/catalogs/catalog_items/__init__.py]: Required: assert_no_error

**Comment by dajoRH** [comment]
Lint report for commit a6a88bc5d67ac5d3a1e5cda72d5c954f21ecb950:

`cfme/services/catalogs/catalog_items/__init__.py`:
- :red_circle: Line 3:1: **F401** *'time' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/services/catalogs/catalog_items/__init__.py (hunk 2 above)
Comment (quarckster) [reply cfme/services/catalogs/catalog_items/__init__.py]: Done

File: cfme/services/catalogs/catalog_items/__init__.py (hunk 4 above)
Comment (quarckster) [reply cfme/services/catalogs/catalog_items/__init__.py]: Done

**Comment by quarckster** [comment]
Failed PRT tests are not related to the PR changes.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 5]
```diff
@@ -0,0 +1,572 @@
...
+        return (
+            self.in_explorer and
+            self.catalog_items.is_opened and
+            self.title.text == 'Adding a new Service Catalog Item'
+        )
+
+
+class AddCatalogItemView(BasicInfoForm):
+    """NonCloudInfraCatalogItem catalog items have this view."""
+    add_button = Button('Add')
```
Comment (sshveta) [inline cfme/services/catalogs/catalog_items/__init__.py]: Please remove _button suffix

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 6]
```diff
@@ -0,0 +1,572 @@
...
+            self.in_explorer and
+            self.catalog_items.is_opened and
+            self.title.text == 'Adding a new Service Catalog Item'
+        )
+
+
+class TabbedAddCatalogItemView(ServicesCatalogView):
+    """Cloud and Infra catalog items have this view."""
+    add_button = Button('Add')
+    cancel_button = Button('Cancel')
```
Comment (sshveta) [inline cfme/services/catalogs/catalog_items/__init__.py]: Please remove _button suffix

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 7]
```diff
@@ -0,0 +1,572 @@
...
+            self.in_explorer and
+            self.catalog_items.is_opened and
+            self.title.text == 'Adding a new Service Catalog Item'
+        )
+
+
+class EditCatalogItemView(BasicInfoForm):
+    save_button = Button('Save')
+    reset_button = Button('Reset')
+    cancel_button = Button('Cancel')
```
Comment (sshveta) [inline cfme/services/catalogs/catalog_items/__init__.py]: Please remove _button suffix

File: cfme/services/catalogs/catalog_items/__init__.py [hunk 8]
```diff
@@ -0,0 +1,572 @@
...
+    def after_fill(self, was_change):
+        # TODO: This is a workaround (Jira RHCFQE-5429)
+        if was_change:
+            wait_for(lambda: not self.save_button.disabled, timeout='10s', delay=0.2)
+
+
+class TabbedEditCatalogItemView(ServicesCatalogView):
+    save_button = Button('Save')
+    reset_button = Button('Reset')
+    cancel_button = Button('Cancel')
```
Comment (sshveta) [inline cfme/services/catalogs/catalog_items/__init__.py]: Please remove _button suffix

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by psav** [comment]
Sorry @quarckster a rebase required here

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by quarckster** [comment]
PRT failures are not related to the changes.

**Pull request merged**